### PR TITLE
Add deletion_protection field to Datafusion Instance

### PR DESCRIPTION
**Pull request by SamanthaMathews**
Add deletion_protection field to make deletion actions require an explicit intent

Part of https://github.com/hashicorp/terraform-provider-google/issues/18854

**Release Note Template for Downstream PRs (will be copied)**
```

Datafusion: added `deletion_protection` field to `datafusion_instance` to make deleting them require an explicit intent. `datafusion_instance` resources now cannot be destroyed unless `deletion_protection = false` is set for the resource.
```

### CODE REVIEW

File: mmv1/products/datafusion/Instance.yaml [hunk 1]
```diff
@@ -35,6 +35,17 @@ timeouts: !ruby/object:Api::Timeouts
   update_minutes: 25
   delete_minutes: 50
 autogen_async: true
+virtual_fields:
+  - !ruby/object:Api::Type::Boolean
+    name: 'deletion_protection'
+    default_value: false
+    description: |
+      Whether Terraform will be prevented from destroying the instance. Defaults to true.
```
Comment (abd-goog) [inline mmv1/products/datafusion/Instance.yaml:43]: Defaults to false.

File: mmv1/products/datafusion/Instance.yaml [hunk 2]
```diff
@@ -35,6 +35,17 @@ timeouts: !ruby/object:Api::Timeouts
...
+      When a`terraform destroy` or `terraform apply` would delete the instance,
+      the command will fail if this field is not set to false in Terraform state.
```
Comment (abd-goog) [inline mmv1/products/datafusion/Instance.yaml:45]: Remove this line as below lines would cover the behaviour

File: mmv1/products/datafusion/Instance.yaml [hunk 3]
```diff
@@ -35,6 +35,17 @@ timeouts: !ruby/object:Api::Timeouts
...
+      When the field is set to true or unset in Terraform state, a `terraform apply`
+      or `terraform destroy` that would delete the instance will fail.
```
Comment (abd-goog) [inline mmv1/products/datafusion/Instance.yaml:47]: When the field is set to true in Terraform state, a `terraform apply` or `terraform destroy` that would delete the instance will fail. When the field is set to false or unset, deleting the instance is allowed

File: mmv1/products/datafusion/go_Instance.yaml [hunk 4]
```diff
@@ -47,9 +47,21 @@ iam_policy:
   import_format:
     - 'projects/{{project}}/locations/{{location}}/instances/{{name}}'
     - '{{name}}'
+virtual_fields:
+  - !ruby/object:Api::Type::Boolean
+    name: 'deletion_protection'
+    default_value: false
+    description: |
+      Whether Terraform will be prevented from destroying the instance. Defaults to true.
```
Comment (abd-goog) [inline mmv1/products/datafusion/go_Instance.yaml:55]: Ditto. Update description to consider the default value is false and not true.

File: mmv1/templates/terraform/pre_delete/datafusion_instance.go.erb [hunk 5]
```diff
@@ -0,0 +1,3 @@
+if d.Get("deletion_protection").(bool) {
+    return fmt.Errorf("cannot destroy domain without setting deletion_protection=false and running `terraform apply`")
```
Comment (abd-goog) [inline mmv1/templates/terraform/pre_delete/datafusion_instance.go.erb:2]: cannot destroy datafusion instance

File: mmv1/templates/terraform/pre_delete/go/datafusion_instance.go.tmpl [hunk 6]
```diff
@@ -0,0 +1,3 @@
+if d.Get("deletion_protection").(bool) {
+    return fmt.Errorf("cannot destroy service without setting deletion_protection=false and running `terraform apply`")
```
Comment (abd-goog) [inline mmv1/templates/terraform/pre_delete/go/datafusion_instance.go.tmpl:2]: cannot destroy datafusion instance.

File: mmv1/third_party/terraform/website/docs/guides/version_6_upgrade.html.markdown [hunk 7]
```diff
@@ -414,4 +414,17 @@ that are derived from the API.
 
 ### `host.gce_instance.disable_ssh` now defaults to true
 
-`disable_ssh` field now defaults to true. To enable SSH, please set `disable_ssh` to false.
\ No newline at end of file
+`disable_ssh` field now defaults to true. To enable SSH, please set `disable_ssh` to false.
+
```
Comment (abd-goog) [inline mmv1/third_party/terraform/website/docs/guides/version_6_upgrade.html.markdown:418]: version 6 is already released. No release notes needed here.

File: mmv1/third_party/terraform/services/datafusion/resource_data_fusion_instance_test.go [hunk 8]
```diff
@@ -123,6 +126,7 @@ resource "google_data_fusion_instance" "foobar" {
   name   = "%s"
   region = "us-central1"
   type   = "ENTERPRISE"
+  deletion_protection = false
```
Comment (abd-goog) [inline mmv1/third_party/terraform/services/datafusion/resource_data_fusion_instance_test.go:129]: no need to set.... it is by default false.

File: mmv1/third_party/terraform/services/datafusion/resource_data_fusion_instance_test.go [hunk 9]
```diff
@@ -140,6 +144,7 @@ resource "google_data_fusion_instance" "foobar" {
   enable_stackdriver_monitoring = true
   enable_stackdriver_logging    = true
   enable_rbac                   = true
+  deletion_protection = false
```
Comment (abd-goog) [inline mmv1/third_party/terraform/services/datafusion/resource_data_fusion_instance_test.go:147]: Ditto

File: mmv1/third_party/terraform/services/datafusion/resource_data_fusion_instance_test.go [hunk 10]
```diff
@@ -207,3 +213,40 @@ resource "google_data_fusion_instance" "basic_instance" {
 }
 `, context)
 }
+
```
Comment (abd-goog) [inline mmv1/third_party/terraform/services/datafusion/resource_data_fusion_instance_test.go:216]: This test is basically not doing anything.
We should be testing creation with deletionProtection true and then the delete fails with an error.

**Review by abd-goog — CHANGES_REQUESTED** [review]

File: mmv1/products/datafusion/Instance.yaml (hunk 1 above)
Comment (SamanthaMathews) [reply mmv1/products/datafusion/Instance.yaml:43]: done

File: mmv1/products/datafusion/Instance.yaml (hunk 2 above)
Comment (SamanthaMathews) [reply mmv1/products/datafusion/Instance.yaml:45]: done

File: mmv1/products/datafusion/Instance.yaml (hunk 3 above)
Comment (SamanthaMathews) [reply mmv1/products/datafusion/Instance.yaml:47]: done

File: mmv1/products/datafusion/go_Instance.yaml (hunk 4 above)
Comment (SamanthaMathews) [reply mmv1/products/datafusion/go_Instance.yaml:55]: done

File: mmv1/templates/terraform/pre_delete/datafusion_instance.go.erb (hunk 5 above)
Comment (SamanthaMathews) [reply mmv1/templates/terraform/pre_delete/datafusion_instance.go.erb:2]: done

File: mmv1/third_party/terraform/services/datafusion/resource_data_fusion_instance_test.go (hunk 8 above)
Comment (SamanthaMathews) [reply mmv1/third_party/terraform/services/datafusion/resource_data_fusion_instance_test.go:129]: done

File: mmv1/third_party/terraform/services/datafusion/resource_data_fusion_instance_test.go (hunk 9 above)
Comment (SamanthaMathews) [reply mmv1/third_party/terraform/services/datafusion/resource_data_fusion_instance_test.go:147]: done

File: mmv1/third_party/terraform/services/datafusion/resource_data_fusion_instance_test.go (hunk 10 above)
Comment (SamanthaMathews) [reply mmv1/third_party/terraform/services/datafusion/resource_data_fusion_instance_test.go:216]: done

File: mmv1/products/datafusion/go_Instance.yaml [hunk 11]
```diff
@@ -47,9 +47,18 @@ iam_policy:
   import_format:
     - 'projects/{{project}}/locations/{{location}}/instances/{{name}}'
     - '{{name}}'
-custom_code:
```
Comment (abd-goog) [inline mmv1/products/datafusion/go_Instance.yaml:50]: why remove custom_code? Won't work without this.

File: mmv1/third_party/terraform/website/docs/guides/version_6_upgrade.html.markdown [hunk 12]
```diff
@@ -413,5 +413,3 @@ that are derived from the API.
 ## Resource: `google_workstations_workstation_config`
 
 ### `host.gce_instance.disable_ssh` now defaults to true
-
```
Comment (abd-goog) [inline mmv1/third_party/terraform/website/docs/guides/version_6_upgrade.html.markdown:416]: looks like unintended removal of disable_ssh. Please leave this file untouched.

File: mmv1/third_party/terraform/services/datafusion/resource_data_fusion_instance_test.go [hunk 13]
```diff
@@ -207,3 +213,40 @@ resource "google_data_fusion_instance" "basic_instance" {
...
+		Steps: []resource.TestStep{
+			{
+				Config: testAccDatafusionInstanceDeletionProtection("us-central1"),
+			},
+			{
+				ResourceName:            "google_data_fusion_instance.instance",
+				ImportState:             true,
+				ImportStateVerify:       true,
+				ImportStateVerifyIgnore: []string{"labels", "terraform_labels", "deletion_protection"},
+			},
```
Comment (abd-goog) [inline mmv1/third_party/terraform/services/datafusion/resource_data_fusion_instance_test.go:233]: Can you try running this test locally? Mostly the post test destroy test should fail. 
To fix that we need to add another step that sets deletion_protection = false after the test.
smth like: {Config: testAccDatafusionInstanceDeletionProtectionFalse("us-central1")}
where
func testAccDatafusionInstanceDeletionProtectionFalse(region string) string {

	return fmt.Sprintf(`
	resource "google_data_fusion_instance" "instance" {
        name   = "my-instance"
        region = "%s"
        type   = "BASIC"
	deletion_protection = false
}
`, region)
}

**Review by abd-goog — CHANGES_REQUESTED** [review]

**Comment by c2thorn** [comment]
Discussing with @SamanthaMathews offline about this PR and the go rewrite migration

**Comment by c2thorn** [comment]
Hi Samantha, please recreate this PR per our conversation earlier this month.

**Pull request closed without merging**